lib/evmone/instructions_calls.cpp

```diff
@@ -50,7 +55,7 @@ evmc_status_code call(ExecutionState& state) noexcept
 
     auto cost = has_value ? 9000 : 0;
 
-    if constexpr (Kind == EVMC_CALL)
+    if constexpr (Op == OP_CALL)
```

Comment:
This looks suspicious, because previously this was CALL or STATICCALL.

I guess for static call conditions inside are always false: `has_value` is false and STATICALL was added after Spurious Dragon...

Maybe makes sense to add some assertions here for these assumptions?

Reply:
Yes. Previously I missed this and there was a bug related to the other check like this. But here this seems correct. I will add asserts. I don't want to change the structure of the implementation but also redundantly adding `Op == OP_STATICCALL` felt wrong.

Reply:
I can add even static asserts.
```patch
Index: lib/evmone/instructions_calls.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lib/evmone/instructions_calls.cpp b/lib/evmone/instructions_calls.cpp
--- a/lib/evmone/instructions_calls.cpp	(revision 078fe563dffa80e2c298080a8202f4e87434d1a0)
+++ b/lib/evmone/instructions_calls.cpp	(date 1642011381658)
@@ -12,9 +12,11 @@
     static_assert(
         Op == OP_CALL || Op == OP_CALLCODE || Op == OP_DELEGATECALL || Op == OP_STATICCALL);
 
+    constexpr auto has_value_arg = (Op == OP_CALL || Op == OP_CALLCODE);
+
     const auto gas = state.stack.pop();
     const auto dst = intx::be::trunc<evmc::address>(state.stack.pop());
-    const auto value = (Op == OP_STATICCALL || Op == OP_DELEGATECALL) ? 0 : state.stack.pop();
+    const auto value = has_value_arg ? state.stack.pop() : 0;
     const auto has_value = value != 0;
     const auto input_offset = state.stack.pop();
     const auto input_size = state.stack.pop();
@@ -55,6 +57,9 @@
 
     auto cost = has_value ? 9000 : 0;
 
+    // STATICCALL was introduces after Spurious Dragon and never has value.
+    static_assert(instr::traits[OP_STATICCALL].since >= EVMC_SPURIOUS_DRAGON);
+    static_assert(Op != OP_STATICCALL || !has_value_arg);
     if constexpr (Op == OP_CALL)
     {
         if (has_value && state.msg->flags & EVMC_STATIC)
```

But I think these suggest some kind of spec optimization. The core of the problem is that the previous check was too broad — `STATICCALL` was never meant to be included.

First check — `STATICCALL` does not violate static mode.

Second check — this refers to for `CALL`
![image](https://user-images.githubusercontent.com/573380/149200637-009ff3a0-2aaf-4795-96c9-2142ae3047bf.png)
but for `STATICCALL`
![image](https://user-images.githubusercontent.com/573380/149200932-fba930f0-86d9-42b1-b242-6db08f9a8fcf.png)

Reply:
Yeah makes sense :+1: